torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,258 @@
+      registry_;
+  std::mutex lock_;
+};
+
+const std::unordered_map<std::string, c10d::ReduceOp> str_to_reduce_op = {
+    {"sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::SUM)},
+    {"avg", c10d::ReduceOp(c10d::ReduceOp::RedOpType::AVG)},
+    {"product", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PRODUCT)},
+    {"min", c10d::ReduceOp(c10d::ReduceOp::RedOpType::MIN)},
+    {"max", c10d::ReduceOp(c10d::ReduceOp::RedOpType::MAX)},
+    {"band", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BAND)},
+    {"bor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BOR)},
+    {"bxor", c10d::ReduceOp(c10d::ReduceOp::RedOpType::BXOR)},
+    // TODO: support premul_sum
+    // {"premul_sum", c10d::ReduceOp(c10d::ReduceOp::RedOpType::PREMUL_SUM)},
+    {"unused", c10d::ReduceOp(c10d::ReduceOp::RedOpType::UNUSED)}};
+
+c10d::ReduceOp to_reduce_op(const std::string& reduce_op) {
+  auto it = str_to_reduce_op.find(reduce_op);
+  TORCH_CHECK(
+      it != str_to_reduce_op.end(), "Unrecognized reduce_op: ", reduce_op);
+  return it->second;
+}
+
+at::Tensor all_reduce_(
+    at::Tensor input,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  c10d::AllreduceOptions opts;
+  opts.reduceOp = to_reduce_op(reduce_op);
+
+  std::vector<at::Tensor> inputs{input};
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->allreduce(inputs, opts);
+  c10d::RankLocal<WorkRegistry>::get().register_work(input, work);
+  return input;
+}
+
+at::Tensor all_reduce(
+    const at::Tensor& input,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  auto output = input.clone();
+  return all_reduce_(output, reduce_op, group_name);
+}
+
+std::vector<at::Tensor> all_reduce_coalesced_(
+    std::vector<at::Tensor> inputs,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  c10d::AllreduceCoalescedOptions opts;
+  opts.reduceOp = to_reduce_op(reduce_op);
+
+  auto group = c10d::resolve_process_group(group_name);
+  auto work = group->allreduce_coalesced(inputs, opts);
+  for (const auto& tensor : inputs) {
+    c10d::RankLocal<WorkRegistry>::get().register_work(tensor, work);
+  }
+  return inputs;
+}
+
+std::vector<at::Tensor> all_reduce_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const std::string& reduce_op,
+    const std::string& group_name) {
+  std::vector<at::Tensor> outputs;
+  for (const auto& tensor : inputs) {
+    outputs.push_back(tensor.clone());
+  }
+  return all_reduce_coalesced_(outputs, reduce_op, group_name);
+}
+
+at::Tensor allocate_all_gather_output(
+    const at::Tensor& input,
+    int64_t group_size) {
+  auto output_size = input.sizes().vec();
+  output_size[0] *= group_size;
+  return at::empty(
+      output_size,
+      at::TensorOptions().dtype(input.dtype()).device(input.device()));
+}
+
+std::vector<at::Tensor> all_gather_into_tensor_coalesced(
+    const std::vector<at::Tensor>& inputs,
+    const int64_t group_size,
```

Comment:
curious why we need `group_size` as a argument, i thought the `group_size` can be resolved after the we call `group  = resolve_process_group` and then use `group.size()`?

Reply:
I also really don't like the redundancy here.

Since the output shape depends on `group_size`, without it we'd need to query the group info inside meta kernels. This is fine for just-in-time compilation where the group info is available. However, it wouldn't work if we export the program and compile it in another environment.

Reply:
I see, hmmm if we export the program and compile it in another environment, we still need to somehow get the corresponding `group_name -> pg` setup otherwise when we run it in another environment, the process group resolution would also fail?

Reply:
If we are only compiling I think we can get away with not having the mapping available (e.g. model transformation + aot_compile for inference). But yeah we need this information eventually at runtime. I'll think about if we want to export this information in a standard way (and use it to create fake pgs for compilation). Will revisit this before making the API public.

Reply:
Yeah sounds good, let's do it as a follow up. I feel if this is just for the meta function we should be able to hide this inside the meta function/real function op itself by resolving the group and get the group size, as no matter where we run the compiled program we would need to setup some pg.